src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/ConcurrentQueueSemaphore.NetStandard.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClient.SNI
+{
+    internal sealed partial class ConcurrentQueueSemaphore
+    {
+        public Task WaitAsync(CancellationToken cancellationToken)
+        {
+            // try sync wait with 0 which will not block to see if we need to do an async wait
+            if (_semaphore.Wait(0, cancellationToken))
+            {
+                return Task.CompletedTask;
+            }
+            else
+            {
+                var tcs = new TaskCompletionSource<bool>();
+                _queue.Enqueue(tcs);
+                _semaphore.WaitAsync().ContinueWith(
+                    continuationAction: s_continuePop,
+                    state: _queue,
+                    cancellationToken: cancellationToken
```

Comment:
Doesn't it mean that it's going to cancel the continuation on `_semaphore.WaitAsync()`, so there will be no one to dequeue the tcs, and the tcs in question is never going to be cancelled?

Reply:
Unknown, I'm not adding this functionality just making it faster, you probably need @cheenamalhotra

Reply:
It could be addressed, as it's important to end the wait on cancellation than to cancel the deque. The deque should ideally happen even on cancellation.